pkg/client/results/junit.go

```diff
@@ -45,84 +128,126 @@ func Filter(predicate func(testCase reporters.JUnitTestCase) bool, testSuite rep
+
 		return resultObj, errors.Wrapf(err, "opening file %v", currentFile)
 	}
 	defer infile.Close()
 
-	decoder := xml.NewDecoder(infile)
-	if err := decoder.Decode(&junitResults); err != nil {
-		resultObj.Status = StatusUnknown
-		resultObj.Metadata["error"] = err.Error()
-		return resultObj, errors.Wrap(err, "error decoding json into object")
+	resultObj, err = junitProcessReader(
+		infile,
+		resultObj.Name,
+		resultObj.Metadata,
+	)
+	if err != nil {
+		return resultObj, errors.Wrap(err, "error processing junit")
+	}
+
+	return resultObj, nil
+}
+
+func junitProcessReader(r io.Reader, name string, metadata map[string]string) (Item, error) {
+	rootItem := Item{
+		Name:     name,
+		Status:   StatusPassed,
+		Metadata: metadata,
 	}
 
-	for _, t := range junitResults.TestCases {
-		status := StatusUnknown
-		switch {
-		case Passed(t):
-			status = StatusPassed
-		case Failed(t):
-			resultObj.Status = StatusFailed
-			status = StatusFailed
-		case Skipped(t):
-			status = StatusSkipped
+	if r == nil {
+		rootItem.Status = StatusUnknown
+		if rootItem.Metadata == nil {
+			rootItem.Metadata = map[string]string{}
 		}
-		testItem := Item{Name: t.Name, Status: status}
+		rootItem.Metadata["error"] = "no data source for junit"
+		return rootItem, errors.New("no data source for junit")
+	}
 
-		hasFailureMsg := (t.FailureMessage != nil && t.FailureMessage.Message != "")
-		if hasFailureMsg || t.SystemOut != "" {
-			testItem.Details = map[string]string{}
+	decoder := xml.NewDecoder(r)
+	junitResults := junitResult{}
+	if err := decoder.Decode(&junitResults); err != nil {
+		rootItem.Status = StatusUnknown
+		if rootItem.Metadata == nil {
+			rootItem.Metadata = map[string]string{}
 		}
-		if hasFailureMsg {
-			testItem.Details[JunitFailureKey] = t.FailureMessage.Message
+		rootItem.Metadata["error"] = err.Error()
+		return rootItem, errors.Wrap(err, "decoding junit")
+	}
+
+	for _, ts := range junitResults.suites.Suites {
+		suiteItem := Item{
+			Name:   ts.Name,
+			Status: StatusPassed,
 		}
-		if t.SystemOut != "" {
-			testItem.Details[JunitStdoutKey] = t.SystemOut
+		for _, t := range ts.TestCases {
+			status := StatusUnknown
+			switch {
+			case JUnitPassed(t):
+				status = StatusPassed
+			case JUnitFailed(t):
+				rootItem.Status = StatusFailed
+				suiteItem.Status = StatusFailed
+				status = StatusFailed
+			case JUnitSkipped(t):
+				status = StatusSkipped
+			}
+			testItem := Item{Name: t.Name, Status: status, Details: map[string]string{}, Metadata: map[string]string{}}
+
+			// Different JUnit implementations build the objects in slightly different ways.
+			// Some will only use contents, some only the message attribute. Here we just concat
+			// the values, separated with a space.
+			hasFailureContents := (t.Failure != nil && (t.Failure.Message != "" || t.Failure.Contents != ""))
+			if hasFailureContents {
+				testItem.Details[JUnitFailureKey] = strings.TrimSpace(t.Failure.Message + " " + t.Failure.Contents)
```

Comment:
This (and the lines above it) where the most recent change. Just checks that either value exists and joins them (trimming spaces at the end).

Reply:
I think this is fine. You could also build the string up gradually by checking whether either `Message` or `Contents` was empty and appending if they aren't but I don't feel particularly strongly about it. I'm happy for you to leave it as is.